## Functions

Variables are one of the two fundamental components of any programming language.  Now that we know what variables are (they things that store data) and how to work with them, we can introduce the the other fundamental component of programming: functions.

Functions give us the ability to write some code that we reference over and over again without having to retype things.  Ultimately, programming is about making your life easier!  We work hard (say, for one semester of our lives) learning how to give a computer instructions (using Python) and then, for the rest of our lives, everything else is easier and more convenient because we can instruct a computer to do lots of the heavy lifting for us.  <u>In financial terms, we spend a little time now for a bunch of time saved later, and thus the net present value of learning to program is positive.</u>

Functions have three parts:
1. Function defintion (name the function and determine what *inputs* we need, if any)
2. Operations (execute some code)
3. Return statement (define what want we want to get back from the function)

As an example, let's compute the weighted average cost of capital, defined as

$$
r_{\text{WACC}} = \frac{E}{V}\times r_E + \frac{D}{V}\times r_D\times (1-\tau_C).
$$

That is, the weighted average cost of capital for a firm is the proportion of equity in the firm times the equity cost of capital plus the proportion of debt in the firm times the debt cost of capital.  We adjust the latter for the tax shield of debt using the corporate tax rate $\tau_C$.

**Part 1: definition**

A function defintion begins with the keyword `def` followed by the function name and then a colon.

Functions can use information passed to it via input arguments (inputs are named in the parentheses, which appear before the colon).

```python3
def wacc(E, D, rE, rD, tC):
```

Note that in this function definition, we expect `E` and `D` as inputs.  We do not require `V` to be an input to the function, since `V` is easily calculated as `V = E + D`.  This calculation is included in part 2.

**Part 2: operations**

Functions run the code written inside of the function (operations) when the function is *called*.  A function call occurs when a programmer ''uses'' the function, and we'll see an example of this shortly.

<u>NOTE</u>: indentation keeps track of whether code is ''inside'' a function

```python3
def wacc(E, D, rE, rD, tC):
    V = E + D
    cost_of_capital = E / V * rE + D / V * rD * (1 - tC)
```

**Part 3: return value**

Functions can be many, many lines of code.

The last line, starting with the keyword `return` specifies what value comes back when you run the function.

```python3
def wacc(E, D, rE, rD, tC):
    V = E + D
    cost_of_capital = E / V * rE + D / V * rD * (1 - tC)
    return cost_of_capital
```

When a function is created, Python becomes ''aware'' of the function, but nothing else happens at that time.

In [1]:
def wacc(E, D, rE, rD, tC):
    V = E + D
    cost_of_capital = E / V * rE + D / V * rD * (1-tC)
    return cost_of_capital

For functions to be used, they need to be *called*.  For example:

In [2]:
wacc(100,30,.08,.04,.35)

0.06753846153846155

tells Python to compute the weighted average cost of capital for a firm with an equity value of $100$, a debt value of $30$, an equity cost of capital of $8\%$, a debt cost of capital of $4\%$, and a tax rate of $35\%$.  The returned value of `0.0675` tells us that the firm's WACC is $6.75\%$.

In the above example, Python *assumes* the first number ($100$) should be given to the input variable `E`, it *assumes* that the second number ($30$) should be given to the input variable `D`, etc.  This is bad practice!  It's much better to use the input variable names:

In [3]:
wacc(E=100,D=30,rE=.08,rD=.04,tC=.35)

0.06753846153846155

because then the ordering of how you enter the inputs doesn't matter.

In [4]:
wacc(rE=.08,rD=.04,tC=.35,E=100,D=30)

0.06753846153846155

This is useful to keep in mind because, realistically, you won't remember the default ordering of inputs (the order of input variables in the `def` line of the function definition).

The `return` line is important because variables that live inside a function are stuck inside that function.

E.g. the below will throw an error because `cost_of_capital` is only *accessible* inside the function `wacc`.

In [5]:
cost_of_capital

NameError: name 'cost_of_capital' is not defined

In order to get the value `cost_of_capital` back from the function, we specify that this value should be *returned* from the function.  If desired, we can instruct the function to return multiple items to us.  For instance, the `wacc()` function defined above returns only the cost of capital.  If we want to return the total firm value, `V`, along with the cost of capital, we can easily modify the function to do so.

In [6]:
def more_wacc(E, D, rE, rD, tC):
    V = E + D
    cost_of_capital = E / V * rE + D / V * rD * (1-tC)
    return cost_of_capital, V

In [7]:
more_wacc(E=100,D=30,rE=.08,rD=.04,tC=.35)

(0.06753846153846155, 130)

By default, Python is printing out the returned item(s) from the function.  This happens because, if Python is not told to save the returned item(s) anywhere, Python simply prints it out.  We can instead save the returned item(s) to a variable.

In [8]:
r_W = wacc(E=100,D=30,rE=.08,rD=.04,tC=.35)

This will preserve the returned item for later use.  For instance, suppose that we calculate the cost of capital and then estimate the present value of a series of three dividend payments (one today, one in a year from now, and one in two years from now).  For simplicity, let's assume each divident payment is \$5.

In [9]:
5 + 5/(1+r_W) + 5/(1+r_W)**2

14.071028677403966

If multiple items are returned from a function and we save the returned items to a variable, that variable will be a list.

In [10]:
x = more_wacc(E=100,D=30,rE=.08,rD=.04,tC=.35)
print(x)

(0.06753846153846155, 130)


In [11]:
x[0]

0.06753846153846155

<span style="color:red">**Concept check**:</span>  Use the `more_wacc()` function to calculate WACC under the following parameter values:
- E=200
- D=80
- rE=.09
- rD=0.8
- tC=.4

In [ ]:
# complete concept check here

### Plan Ahead

Functions have the potential to create immense utility for the end user.  A very well constructed function can save you hours (if not days or weeks) of time down the road.  However, to be useful, the function needs to be planned carefully.  If the function is written well, then you can use it conveniently in the future without worrying about the details.  However, if the function is written poorly, it could end up creating a huge headache for you when you try to use it later.

Keeping to the `wacc()` example, let's consider some simple improvements to the function.

Imagine that we plan to use this `wacc()` function to value companies, using data on companies that will be obtained in some other piece of code.  Thus, we collect all the requisite components to calculating a weighted average cost of capital, as well as information about the current free cash flow and an expected growth rate to the free cash flow.  For simplicity, assume that free cash flow growth and the risk free rate of capital are constant.  In this scenario, we can apply the Gordon growth formula to value the company.

$$
V = \frac{FCF}{r_{WACC}-g}
$$

In [18]:
2/(wacc(E=100,D=30,rE=.08,rD=.04,tC=.35) - 0.01) # $2M in FCF, growing at 1%/year

34.75935828877005

A limitation of the `wacc()` function is that it assumes all inputs are sensible.  Suppose an innocent typo places a negative sign in front of the value for `rE`.  In this case, form value can end up a negative number!

In [19]:
2/(wacc(E=100,D=30,rE=-.08,rD=.04,tC=.35) - 0.01)

-30.516431924882625

One can look at this and see that there is a problem; firm value cannot be negative.  A scarier example is one in which the accidental negative shows up in front of the value for `rD`.  This latter case is worse because it's not as obvious that there is an issue (firm value is still calculated to be positive).

In [20]:
2/(wacc(E=100,D=30,rE=.08,rD=-.04,tC=.35) - 0.01)

43.91891891891892

These sorts of small problems can create huge issues for you or your company.  In this working example, the typo over-values the firm by:

In [23]:
value_wrong = 2/(wacc(E=100,D=30,rE=.08,rD=-.04,tC=.35) - 0.01)
value_right = 2/(wacc(E=100,D=30,rE=.08,rD=.04,tC=.35) - 0.01)
value_wrong / value_right - 1

0.2635135135135136

Therefore, when we write the `wacc()` function, we should take care to watch out for anticipable errors and deal with them.

One option is to simply print out a warning message if the function detects something fishy.

In [26]:
def better_wacc(E, D, rE, rD, tC):
    
    if rE <= 0:
        print('rE is not positive')
    if rD <= 0:
        print('rD is not positive')
    if tC <= 0:
        print('tC is not positive')
    
    V = E + D
    cost_of_capital = E / V * rE + D / V * rD * (1-tC)
    return cost_of_capital

2/(better_wacc(E=100,D=30,rE=.08,rD=-.04,tC=.35) - 0.01)

Warning! rD is not positive


43.91891891891892

This at least generates a warning message, but warnings can only go so far.  There is a very real possibility that someone using this function may not notice the warning message and consequently carry on with their analysis despite having an erroneously calculated firm value.

If you are certain something is absolutely not an acceptable input, stop it!  Do not let the function continue operating on bad data.  The way to do this is with an **Exception**.

In [27]:
def better_wacc(E, D, rE, rD, tC):
    
    if rE <= 0:
        raise Exception('rE is not positive')
    if rD <= 0:
        raise Exception('rD is not positive')
    if tC <= 0:
        raise Exception('tC is not positive')
    
    V = E + D
    cost_of_capital = E / V * rE + D / V * rD * (1-tC)
    return cost_of_capital

2/(better_wacc(E=100,D=30,rE=.08,rD=-.04,tC=.35) - 0.01)

Exception: Warning! rD is not positive

Exception messages spit out lots of text, so they can be somewhat intimidating to approach at first.  Python will print out a *traceback* of what code ran that ultimately resulted in an error.  The easiest way to get a quick picture of what happened is to follow the `--->` symbols, reading from top to bottom.  The first such symbol indicates that line 14 was called.  The second such symbol tells us that line 6 was called.  The actual error message is the final line of the printout.  From this brief inspection, we can figure out what went wrong.  Notice that line 6 can only be called if line 5 is true.  Hence, the root of our problem is that line 5 is true, indicating that `rD` is entered as a negative number.

The `if` statements added to `better_wacc()` cause the funcion to "break" (raise an error message and stop the code from continuing) under certain scenarios.  To see that the code completely terminantes when an exception is raised, try the following.

In [28]:
print( 2/(better_wacc(E=100,D=30,rE=.08,rD=-.04,tC=.35) - 0.01) )
print('firm value is now calculated')

Exception: Warning! rD is not positive

Note that the second print statement is not executed.  Python halts all operations when an exception occurs.

Do not insert exceptions to your functions unnecessarily.  Only rely on financial principles to make these judgements.  It is obvious, for example, that costs of capital should not be negative.  Moreover, the second Modigliani and Miller proposition tells us that `rE` should always be at least as high as `rD`, so we could raise an exception if `rE <= rD`.  However, there is no strict rule, either by regulation or by principles of finance, that debt value cannot exceed equity value.  For some smaller firms, it may very well be that leverage exceeds 50%.  However, if you write the `better_wacc()` function and anticipate that it will only be used on large, S\&P 500 companies, it may make sense to check whether `E > D`.  In this case, raising an exception would not be appropriate.  However, since we would generally expect that equity value should exceed debt value, it may make sense to print a warning message for the user.

In [40]:
def better_wacc(E, D, rE, rD, tC):
    
    if rE <= 0:
        raise Exception('rE is not positive')
    if rD <= 0:
        raise Exception('rD is not positive')
    if tC < 0:
        raise Exception('tC is negative')
    if rE <= rD:
        raise Exception('rE <= rD, M&M prop2 violated')
    if D > E:
        print('Warning, debt value exceeds equity value.  Did you accidentally flip these values?')
    
    V = E + D
    cost_of_capital = E / V * rE + D / V * rD * (1-tC)
    return cost_of_capital

<span style="color:red">**Concept check**:</span>  Use the `better_wacc()` function to estimate WACC given the following parameter values.
- E = 10
- D = 100
- rE = .15
- rD = .05
- tC = .35

By the way, does the gap between `rE` and `rD` make senese to you?  Hint: think about the levels of `E` and `D`.

In [ ]:
# complete concept check here

### Combine Multiple Functions for Maximal Efficiency

Functions should handle small, specific tasks.  Do not write one large function that does an entirely specialized task that you will rarely need to perform.  Remember, the goal of functions is to create shortcuts for us to instruct Python to perform a repetitive task, rather than re-writing those instructions each time we need a task to be performed.

You will find that one convenient use of functions is to combine them in a manner like what is shown below.  We already have the `better_wacc()` function defined, which calculates a firm's weighted average cost of capital.  As alluded to above, a firm's cost of equity is a function of its leverage ratio.  So let's create a function called `cost_of_equity()` that will determine a firm's cost of equity capital.  Recall that Modigliani and Miller proposition 2 states:
$$
r_E = r_A + (r_A-r_D)*(1-t_C)*(D/E)
$$
where $r_A$ is the *unlevered* cost of capital (equal to the cost of equity capital for a firm with no leverage).

Let us then define another function, `firm_value()` that calculates the value of a firm.  Inside the `firm_value()` function, we will call both the `better_wacc()` function and the `cost_of_equity()` function.

In [35]:
def cost_of_equity(E, D, rA, rD, tC):
    return rA + (rA-rD)*(1-tC)*(D/E)

def firm_value(E, D, rA, rD, tC, FCF, g):
    
    rE = cost_of_equity(E=E, D=D, rA=rA, rD=rD, tC=tC)
    rW = better_wacc(E, D, rE, rD, tC)
    
    if rW < g:
        raise Exception('rW must be greater than g, but g=', g, 'and rW=', rW)
    
    return FCF / (rW - g)

firm_value(E=100, D=30, rA=.05, rD=.04, tC=.35, FCF=2, g=.01)

55.61497326203209

What happens if the firm had a different debt level?

In [36]:
firm_value(E=90, D=40, rA=.05, rD=.04, tC=.35, FCF=2, g=.01)

57.77777777777777

Debt is valuable because of the tax shield.  You can see that this is the cause for the change in firm value by setting `tC` to zero and re-computing both values above.

In [41]:
print( firm_value(E=100, D=30, rA=.05, rD=.04, tC=0, FCF=2, g=.01) )
print( firm_value(E=90, D=40, rA=.05, rD=.04, tC=0, FCF=2, g=.01) )

50.0
50.0


<span style="color:red">**Concept check**:</span>  Estimate the cost of equity for a firm under two different scenarios.
1. E=90, D=10, rA=.05, rD=.04, tC=.2
2. E=10, D=90, rA=.05, rD=.04, tC=.2

In [ ]:
# complete concept check here